src/webgpu/api/validation/queue/copyToTexture/CopyExternalImageToTexture.spec.ts

```diff
@@ -405,38 +494,27 @@ g.test('source_offscreenCanvas,state')
   )
   .fn(async t => {
     const { detached, copySize } = t.params;
-    const offscreenCanvas = t.getOffscreenCanvas(1, 1);
+    const offscreenCanvas = createOffscreenCanvas(t, 1, 1);
     const dstTexture = t.device.createTexture({
       size: { width: 1, height: 1, depthOrArrayLayers: 1 },
       format: 'bgra8unorm',
       usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.RENDER_ATTACHMENT,
     });
 
     if (detached) {
-      const workerCode =
-        "self.onmessage = function(e) {e.data.canvas.getContext('2d'); poseMessage('get 2d context');};";
-      const blob = new Blob([workerCode], { type: 'text/javascript' });
-      const url = URL.createObjectURL(blob);
-      const worker = new Worker(url);
-      worker.postMessage({ canvas: offscreenCanvas }, [offscreenCanvas]);
-      worker.onmessage = function (e) {
-        t.runTest(
-          { source: offscreenCanvas },
-          { texture: dstTexture },
-          copySize,
-          true, // No validation errors.
-          'InvalidStateError'
-        );
-      };
+      const messageChannel = new MessageChannel();
+      messageChannel.port1.postMessage(offscreenCanvas, [offscreenCanvas]);
     } else {
       offscreenCanvas.getContext('2d');
```

Comment:
Same concern here. Would also be solved by adding a case which creates a context on the transferred canvas (after waiting for it to arrive on port2).

Reply:
@kainino0x Yes, but I hit an issue when I tried to add a case which creates a context on the transferred canvas. I'll list this in TODO but left comments here so maybe you could help, here is the issue:
- Register onmessage function for port2 and in the onmessage callback, we do `getContext()` and postMessage
- Register onmessage function for port1 and in the onmessage callback we do `t.runTest`.
- Call `messageChannel.port1.postMessage(offscreenCanvas, [offscreenCanvas])`.
But this model doesn't works because when the port1 onmessage callback has been triggered, the Fixture class already return the webgpu device to the pool. With some debugging work, I find that after executing the `messageChannel.port1.postMessage`, case will hit the end of the closure, and this assert(https://github.com/gpuweb/cts/blob/main/src/webgpu/gpu_test.ts#L49) fails when running `t.runTest` in the `port1.onmessage`.
I'm not quite sure whether current cts framework can wait for the callback and then terminate.

Reply:
I think you would need something like:

```ts
const port2FirstMessage = new Promise(resolve => {
  messageChannel.port2.onmessage = m => resolve(m);
});

messageChannel.port1.postMessage(offscreenCanvas, [offscreenCanvas]);

const receivedOffscreenCanvas = await port2FirstMessage;
```

Reply:
Wow, that's quite clever!  And seems work.